### PR TITLE
[RLlib; Offline RL] Enable offline training without specifying an environment.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Offline RL trains an agent on an offline dataset. In RLlib a user still needs to provide an environment when training on offline data which is a hard restriction. There are many cases in which there might not even exist an environment - next to an obvious reason to use offline data: because sampling is expensive:
- The data was recorded by a physical device or facility
- The data was recorded by humans themselves

This PR proposes running Offline RL in RLlib without specifying an environment in the configuration. A user has to provide 
- An action space to be used for the `RLModule`
- An observation space to be used for the `RLModule`.
The algorithm runs - in case of no evaluation - without any `EnvRunnerGroup` and updates modules where needed directly from the `LearnerGroup` instead.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm_config.py [hunk 1]
```diff
@@ -4737,6 +4751,22 @@ def _validate_offline_settings(self):
                 "recorded episodes cannot be read in for training."
             )
 
+    @property
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:4754]: cool!

File: rllib/algorithms/marwil/marwil.py [hunk 2]
```diff
@@ -61,7 +63,15 @@ class MARWILConfig(AlgorithmConfig):
         )
         # Define the environment for which to learn a policy
         # from offline data.
-        config.environment("CartPole-v1")
+        config.environment(
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:66]: Perfect!

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -667,54 +667,44 @@ def setup(self, config: AlgorithmConfig) -> None:
...
-        ):
+        if self.config.is_offline:
             from ray.rllib.offline.offline_data import OfflineData
 
             self.offline_data = OfflineData(self.config)
         # Otherwise set the attribute to `None`.
         else:
             self.offline_data = None
 
+        if not self.offline_data:
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:680]: Should we re-use the same if-block as above (its `else` sub-block)?

File: rllib/algorithms/algorithm.py [hunk 4]
```diff
@@ -667,54 +667,44 @@ def setup(self, config: AlgorithmConfig) -> None:
             )
             self.config.off_policy_estimation_methods = ope_dict
 
-        # Create a set of env runner actors via a EnvRunnerGroup.
-        self.env_runner_group = EnvRunnerGroup(
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:671]: Oh wow, this is kind of cool. So offline algos won't have any `EnvRunnerGroup` anymore at all?!

If all tests pass with this, I'm totally fine with it :D But yeah, should work: The Learner now carries the main RLModule copy, so what was previously the task of the "local" EnvRunner is now not needed anymore.

File: rllib/tuned_examples/ppo/cartpole_ppo.py [hunk 5]
```diff
@@ -25,6 +25,7 @@
     )
 )
 
+args.no_tune = True
```
Comment (sven1977) [inline rllib/tuned_examples/ppo/cartpole_ppo.py:28]: nit: remove this line again.

File: rllib/algorithms/marwil/marwil.py [hunk 6]
```diff
@@ -465,15 +483,16 @@ class (multi-/single-learner setup) and evaluation on
...
-        # Update weights - after learning on the local worker -
-        # on all remote workers.
-        with self.metrics.log_time((TIMERS, SYNCH_WORKER_WEIGHTS_TIMER)):
-            self.env_runner_group.sync_weights(
-                # Sync weights from learner_group to all EnvRunners.
-                from_worker_or_learner_group=self.learner_group,
-                policies=modules_to_update,
-                inference_only=True,
-            )
+        if self.eval_env_runner_group:
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:486]: Nice!!

File: rllib/algorithms/algorithm_config.py [hunk 7]
```diff
@@ -4691,6 +4691,20 @@ def _validate_to_be_deprecated_settings(self):
...
+        # or does not want to run evaluation, she needs to provide at least
+        # action and observation spaces. Note, we require here the spaces,
+        # i.e. a user cannot provide an environment instead because we do
+        # not want to create the environment to receive spaces.
+        if self.is_offline and (
+            not (self.evaluation_num_env_runners > 0 or self.evaluation_interval)
+            and (self.action_space is None or self.observation_space is None)
+        ):
+            raise ValueError(
+                "If no evaluation should be run `action_space` and "
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:4704]:
```suggestion
                "If no evaluation should be run, `action_space` and "
```

File: rllib/algorithms/algorithm.py [hunk 8]
```diff
@@ -3535,18 +3559,21 @@ def _compile_iteration_results_new_api_stack(self, *, train_results, eval_result
...
-                self.env_runner_group.num_remote_worker_restarts()
-            ),
-        }
-        results["env_runner_group"] = {
-            "actor_manager_num_outstanding_async_reqs": (
-                self.env_runner_group.num_in_flight_async_reqs()
-            ),
-        }
+        if self.env_runner_group:
+            # Fault tolerance stats.
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:3563]:
```suggestion
        # EnvRunner actors fault tolerance stats.
        if self.env_runner_group:
```

**Review by sven1977 — APPROVED** [review]
LGTM! Let's try this ... 🤞 
Thanks for this important change @simonsays1980 !

**Comment by sven1977** [comment]
Hey @simonsays1980 , one important question on CQL left.

File: rllib/algorithms/marwil/marwil.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil.py:486]: Great point! Will take a look at it now.

**Pull request merged**